substrate/validator-sets/pallet/src/lib.rs

```diff
@@ -127,6 +187,227 @@ pub mod pallet {
 
       Ok(())
     }
+
+    /// Will set the given account to validator list if the allocated bond is enough.
+    pub fn add_participant(
+      account: T::AccountId,
+      amount: Amount,
+      network: NetworkId,
+    ) -> DispatchResult {
+      let key = ValidatorSet { session: Self::current_session(), network };
+      if amount < Self::validator_set(key).unwrap().bond {
+        Err(Error::<T>::InSufficientBond)?;
+      }
+
+      // Since this participant will be included after the next session, save it for the set
+      // where it will be active.
+      let participant_set =
+        ValidatorSet { session: Session(Self::current_session().0 + 2), network };
+      JoiningValidators::<T>::try_mutate_exists(participant_set, |existing| {
+        if existing.is_some() {
+          existing.as_mut().unwrap().push((account, amount));
+        } else {
+          *existing = Some(vec![(account, amount)]);
+        }
+        Ok::<_, Error<T>>(())
+      })
+      .unwrap();
+
+      Ok(())
+    }
+
+    /// This function will set the participant for leaving if remaining bond
+    /// falls short for the bond requirement of the set after the deallocation.
+    /// Otherwise it will just deallocate the given `amount`.
+    pub fn maybe_remove_participant(
+      account: T::AccountId,
+      amount: Amount,
+      network: NetworkId,
+    ) -> DispatchResult {
+      let current_set =
+        Self::validator_set(ValidatorSet { session: Self::current_session(), network }).unwrap();
+      let key = ValidatorSet { session: Session(Self::current_session().0 + 2), network };
+
+      let pair = match current_set.participants.iter().find(|p| p.0 == account) {
+        Some(p) => *p,
+        None => {
+          // check whether it is still in the joining set before getting active
+          let joining_set = JoiningValidators::<T>::get(key);
+          *joining_set.iter().find(|p| p.0 == account).ok_or(Error::<T>::NonExistentValidator)?
+        }
+      };
+
+      // check validator has enough to deallocate
+      if amount > pair.1 {
+        Err(Error::<T>::InSufficientAllocation)?;
+      }
+
+      // if the remaining bond is not enough remove from the validator set and
+      // deallocate all bond instead of just "amount" since they will be leaving.
```

Comment:
I'd error, unless they deallocate the entire amount. Then someone trying to deallocate 10 doesn't undo a 100k bond position.